R/getGenomicTiles.R

```diff
@@ -0,0 +1,180 @@
+#' @importFrom GenomeInfoDb seqlengths
+#' @importFrom Biostrings fasta.seqlengths
+#' @importFrom methods is
+#' @importFrom IRanges overlapsAny
+#' @importFrom S4Vectors queryHits subjectHits
+#' 
+#' @export
+getGenomicTiles <- function(genome,
+                            tileWidth,
+                            hasOverlap = list(),
+                            fracOverlap = list(),
+                            numOverlap = list(),
+                            nearest = list()) {
+    ## check arguments
+    ## ... genome
+    if(is(genome, "BSgenome")) {
+        chrlens <- GenomeInfoDb::seqlengths(BSgenome::seqinfo(genome))
+    } else if (is.character(genome) && length(genome) == 1L) {
+        if (suppressWarnings(require(genome, character.only = TRUE, quietly = TRUE))) {
+            genome <- get(genome)
+            chrlens <- GenomeInfoDb::seqlengths(BSgenome::seqinfo(genome))
+        } else if (file.exists(genome)) {
+            chrlens <- Biostrings::fasta.seqlengths(genome)
+        } else {
+            stop("'genome' is neither a valid file nor a BSgenome object.")
+        }
+    } else if (is.numeric(genome) && !is.null(names(genome))) {
+        chrlens <- genome
+    } else {
+        stop("'genome' is not a valid argument for getGenomicTiles()")
+    }
+    ## ... other arguments
+    stopifnot(exprs = {
+        # tileWidth
+        is.numeric(tileWidth)
+        length(tileWidth) == 1L
+        tileWidth > 0
+        # hasOverlap
+        is.list(hasOverlap)
+        length(hasOverlap) == 0L || !is.null(names(hasOverlap))
+        all(unlist(lapply(hasOverlap, is, "GRanges")))
+        # fracOverlap
+        is.list(fracOverlap)
+        length(fracOverlap) == 0L || !is.null(names(fracOverlap))
+        all(unlist(lapply(fracOverlap, is, "GRanges")))
+        # numOverlap
+        is.list(numOverlap)
+        length(numOverlap) == 0L || !is.null(names(numOverlap))
+        all(unlist(lapply(numOverlap, is, "GRanges")))
+        # nearest
+        is.list(nearest)
+        length(nearest) == 0L || !is.null(names(nearest))
+        all(unlist(lapply(nearest, is, "GRanges")))
+        all(unlist(lapply(nearest, function(x) !is.null(names(x)))))
+    })
+    
+    ## create tiles
+    gr <- GenomicRanges::tileGenome(seqlengths = chrlens, tilewidth = tileWidth,
+                                    cut.last.tile.in.chrom = TRUE)
+    gr <- gr[GenomicRanges::width(gr) == tileWidth]
+    
+    ## annotate tiles
+    df <- S4Vectors::mcols(gr)
+    
+    ## ... hasOverlap
+    for (i in seq_along(hasOverlap)) {
+        nm <- paste0(names(hasOverlap)[i], ".hasOverlap")
+        df[[nm]] <- IRanges::overlapsAny(gr, hasOverlap[[i]], ignore.strand = TRUE)
+    }
+
+    ## ... fracOverlap
+    for (i in seq_along(fracOverlap)) {
+        gr2 <- fracOverlap[[i]]
+        ov <- GenomicRanges::findOverlaps(gr, gr2, ignore.strand = TRUE)
+        tilesov <- pmin(GenomicRanges::end(gr[S4Vectors::queryHits(ov)]),
```

Comment:
Not sure whether this would work directly if we'd allow also `GRangesLists`.

Reply:
I agree, this would have to be rewritten. I think we could easily support `GRangesList` inputs, by flattening them using either `unlist(grl)` or `unlist(range(grl))` (see above), but we would need to decide which behavior is desired.

Reply:
In this case, I think we can just use `unlist(grl)` and things will work.

Reply:
Agreed